### PR TITLE
Feature: adding oauth2 util - jwt

### PR DESCRIPTION
**Pull request by eapg**
* Adding oauth2_util file with:
    * build_client_credentials_token using joken library version 2.6.0.
    * validate token
 * adding unit test for build_client_credentials_token.

### CODE REVIEW

File: ea_restaurant_data_loader/config/dev.exs [hunk 1]
```diff
@@ -55,3 +55,5 @@ config :phoenix, :stacktrace_depth, 20
 
 # Initialize plugs at runtime for faster development compilation
 config :phoenix, :plug_init_mode, :runtime
+
+config :joken, default_signer: "secret"
```
Comment (reynoldmorel) [inline ea_restaurant_data_loader/config/dev.exs:59]: why did you add another `config` instead of using the existing one above?

File: ea_restaurant_data_loader/lib/auth/token.ex [hunk 2]
```diff
@@ -0,0 +1,3 @@
+defmodule EaRestaurantDataLoader.Token do
```
Comment (reynoldmorel) [inline ea_restaurant_data_loader/lib/auth/token.ex:1]: why not to directly use `Joken.Config` and eliminate this file?

File: ea_restaurant_data_loader/lib/utils/oauth2_util.ex [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      "exp" => DateTime.utc_now() |> DateTime.add(exp_time, :second) |> DateTime.to_unix(),
+      "clientName" => client_name,
+      "scopes" => scopes
+    }
+
+    token = Token.generate_and_sign!(extra_claims, signer(secret_key))
+  end
+
+  def validate_token(token, secret_key) do
+    {:ok, claims} = Token.verify_and_validate(token, signer(secret_key))
```
Comment (reynoldmorel) [inline ea_restaurant_data_loader/lib/utils/oauth2_util.ex]: wondering if it is better to create a `decode_token/2` function? Can you check how do you handle the validate method in the other projects?

File: ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+defmodule EaRestaurantDataLoader.Oauth2UtilTest do
+  use ExUnit.Case
+  doctest EaRestaurantDataLoader.Oauth2Util
+  alias EaRestaurantDataLoader.Oauth2Util
+
+  test "access token" do
```
Comment (reynoldmorel) [inline ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs]: Can you put this within a `describe`?

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: ea_restaurant_data_loader/config/dev.exs (hunk 1 above)
Comment (eapg) [reply ea_restaurant_data_loader/config/dev.exs:59]: This file have many config statements, as joken is not related with the configs above i added separately.

File: ea_restaurant_data_loader/lib/auth/token.ex (hunk 2 above)
Comment (eapg) [reply ea_restaurant_data_loader/lib/auth/token.ex:1]: This is the way how library recommend the use of Joken.Config, This way it will implement some functions for you.

https://hexdocs.pm/joken/Joken.Config.html

File: ea_restaurant_data_loader/lib/utils/oauth2_util.ex (hunk 3 above)
Comment (eapg) [reply ea_restaurant_data_loader/lib/utils/oauth2_util.ex]: This is a similiar aprouch as python where same function validate and decode.

Capture

File: ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs (hunk 4 above)
Comment (eapg) [reply ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs]: Done.

File: ea_restaurant_data_loader/config/dev.exs (hunk 1 above)
Comment (reynoldmorel) [reply ea_restaurant_data_loader/config/dev.exs:59]: Makes sense to me

File: ea_restaurant_data_loader/lib/auth/token.ex (hunk 2 above)
Comment (reynoldmorel) [reply ea_restaurant_data_loader/lib/auth/token.ex:1]: Fair enough!

File: ea_restaurant_data_loader/lib/utils/oauth2_util.ex (hunk 3 above)
Comment (reynoldmorel) [reply ea_restaurant_data_loader/lib/utils/oauth2_util.ex]: yeah but looks like we don't return anything from the python code

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: ea_restaurant_data_loader/lib/utils/oauth2_util.ex (hunk 3 above)
Comment (eapg) [reply ea_restaurant_data_loader/lib/utils/oauth2_util.ex]: Looking deeper in the python project in the moment where i need the claims i call `jwt.decode()` directly. but in this case is like create two functions with same commands, one that return the claims and other that don't return anything, only throw an error when the token is not valid. I think that create two functions that call `Token.verify_and_validate()` is redundant.

Comment (eapg) [reply ea_restaurant_data_loader/lib/utils/oauth2_util.ex]: Done.

File: ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+      token = Oauth2Util.build_client_credentials_token("ea_restaurant", "READ", 1, "1234")
+      {_, claims} = Oauth2Util.get_token_decoded(token, "1234")
+      assert "ea_restaurant" == claims["clientName"]
+      assert "READ" == claims["scopes"]
+    end
+
+    test " verify expired token" do
+      token = Oauth2Util.build_client_credentials_token("ea_restaurant", "READ", -1, "1234")
+      {_, claims} = Oauth2Util.validate_token(token, "1234")
+      [head | tail] = claims
```
Comment (reynoldmorel) [inline ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs]: I think you don't need the `tail`

File: ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    test "build client credential token successfully" do
+      token = Oauth2Util.build_client_credentials_token("ea_restaurant", "READ", 1, "1234")
+      {_, claims} = Oauth2Util.get_token_decoded(token, "1234")
+      assert "ea_restaurant" == claims["clientName"]
+      assert "READ" == claims["scopes"]
+    end
+
+    test " verify expired token" do
+      token = Oauth2Util.build_client_credentials_token("ea_restaurant", "READ", -1, "1234")
+      {_, claims} = Oauth2Util.validate_token(token, "1234")
```
Comment (reynoldmorel) [inline ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs]: isn't this a `map`?

File: ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+      [head | tail] = claims
+      assert elem(head, 1) == "Invalid token"
```
Comment (reynoldmorel) [inline ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs]: is the `head` another array?

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs (hunk 7 above)
Comment (eapg) [reply ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs]: The head is a tuple.

File: ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs (hunk 6 above)
Comment (eapg) [reply ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs]: no, is a tuple with an status and a map inside. ex {status, map}

File: ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs (hunk 5 above)
Comment (eapg) [reply ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs]: i will fix it.

File: ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs (hunk 6 above)
Comment (reynoldmorel) [reply ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs]: but int he previous test is map, could you rename it? seems confusing

File: ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs (hunk 7 above)
Comment (reynoldmorel) [reply ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs]: could you do pattern matching to get the 1 element of the tuple?

File: ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs (hunk 5 above)
Comment (eapg) [reply ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs]: Done.

File: ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs (hunk 6 above)
Comment (eapg) [reply ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs]: My bad, it is a map in both case.

File: ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs (hunk 7 above)
Comment (eapg) [reply ea_restaurant_data_loader/test/lib/utils/oauth2_util_test.exs]: Done.

**Review by reynoldmorel — APPROVED** [review]
